protocol-2.5.md

```diff
@@ -34,10 +37,10 @@ differently.
 
 Sysex-based commands (0x00 - 0x7F) are used for an extended command set.
 
-| type                  | command | first byte          | second byte   | ...            |
-| --------------------- | ------- | ------------------- | ------------- | -------------- |
-| string                | 0x71    | char *string ...    |               |                |
-| firmware name/version | 0x79    | major version       | minor version | char *name ... |
+| type                  | sub-command | first byte          | second byte   | ...            |
```

Comment:
Jeff:  

I finished the first round of development on my Firmata client library (in Java) and have run it with my version of Core Firmata.  It seems to run okay, and so I’m moving on to working on our design for configurability.  My idea is that I’ll implement something and then pass it on to you for consideration.

My design thoughts at the moment:

1.  There’s a distinction between board-level comms and other sorts of features like the scheduler and device drivers for specific devices.  Comms are basic, pin-oriented, enabling capabilities for higher level functions.  As I suggested earlier, I think that the comms capabilities should be “Standard Features”, that is, optional but known to always be available and reliable.

2.  The Core commands and the StandardFeatures are the heart of actually implementing something with Firmata.  I would like to be able to write Features that implement arbitrary functions using the Core and Standard Features.  I think one good way to do it is to make the processSysex capability available to local code as well as remote code on the client.  I think that all we need to do is separate the processing code from the send-to-client code.  So I am going to try splitting the processSysex routines into processSysex and executeSysex.  The execute methods will drop the results into a byte buffer, the process methods will send the bytes back to the client.  If execute is called directly by a device driver, then the results are available in the byte buffer (which I hope will not grow too big!).

3.  With this model, the device drivers will use the same messaging interface that the client uses to communicate with the comms capabilities.  This will facilitate moving device drivers from client to server, among other benefits.

4.  In order to get around the device id problem, one solution is to use a device name in order to open the device and get back a handle (eg, the feature index in FirmataExt), then refer to the handle in all future messages to it.  Thus, we would only need to allocate a small number of sysex sub-codes which would be used by all device drivers.  The device driver sysex messages would be the standard six: open(name), status(handle), control(handle), read(handle), write(handle) and close(handle).  The specific content of the messages will be established by the device driver.  Device name conflicts can be easily resolved by changing the name in one of the device drivers, perhaps even automatically at addFeature time.

So, I’m going to take a crack at this and see how it works out.  I have an I2C temperature device and an I2C DAC to use as target devices.  First I’ll implement a simple device driver on the client side using the I2C messages, then I’ll move that to the server and implement the higher level device driver messages.

More news as events warrant.

Doug